sdks/go/pkg/beam/runners/prism/internal/engine/strategy.go

```diff
@@ -23,16 +23,351 @@ import (
 	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/typex"
 )
 
+// This file is intended to consolidate handling of WindowingStrategy and trigger
+// logic independantly from the bulk of the ElementManager.
+//
+// Triggers by their nature only apply to Aggregation transforms, in particular:
+// GroupByKeys and CoGroupByKeys are aggregations.
+// Triggers also affect downstream side inputs. That is, a side input consumer
+// is vaccuously an aggregation.
+//
+// Triggers are PerKey+PerWindow, and they may or may not use state per trigger.
+//
+// The unique state per trigger is the trickiest bit to handle. In principle
+// it could just be handled by the existing state system, which might be
+// sufficient, but it's not prepared for unique tagging per trigger itself.
+// It would also add additional overhead since state is kept as the serialized
+// bytes, instead of in a manipulatable form.
+//
+// Instead, each key+window state cell contains a trigger specific state map,
+// handled via pointer equality from the trigger itself.
+
 // WinStrat configures the windowing strategy for the stage, based on the
 // stage's input PCollection.
 type WinStrat struct {
 	AllowedLateness time.Duration // Used to extend duration
 }
 
+// EarliestCompletion marks when we can close a window.
 func (ws WinStrat) EarliestCompletion(w typex.Window) mtime.Time {
 	return w.MaxTimestamp().Add(ws.AllowedLateness)
 }
 
 func (ws WinStrat) String() string {
 	return fmt.Sprintf("WinStrat[AllowedLateness:%v]", ws.AllowedLateness)
 }
+
+// triggerInput represents a Key + window + stage's trigger conditions.
+type triggerInput struct {
+	newElementCount    int  // The number of new elements since the last check.
+	endOfWindowReached bool // Whether or not the end of the window has been reached.
+}
+
+// Trigger represents a trigger for a windowing strategy.  A trigger determines when
+// to fire a window based on the arrival of elements and the passage of time.
+//
+// See https://s.apache.org/beam-triggers for a more detailed look at triggers.
+type Trigger interface {
+	// IsReady determines if the trigger is ready to fire based on the current state.
+	isReady(input triggerInput, state *StateData) bool
```

Comment:
`isReady` should be stateless and separated from methods that advance the state (`onElement`, `onFire` etc). See example below in `AfterEach`. This factoring is actually among the most important benefits of moving to the `isReady` model.

Reply:
Done.